scalafix-tests/unit/src/test/scala-2.12/scalafix/tests/cli/ScalafixImplSuite.scala

```diff
@@ -0,0 +1,192 @@
+import scala.meta.io.Classpath
+import scalafix.interfaces.ScalafixDiagnostic
+import scalafix.interfaces.ScalafixException
+import scalafix.interfaces.ScalafixMainCallback
+import scalafix.interfaces.ScalafixMainMode
+import scalafix.internal.reflect.ClasspathOps
+import scalafix.internal.reflect.RuleCompiler
+import scalafix.internal.sbt.ScalafixJarFetcher
+import scalafix.testkit.DiffAssertions
+
+class ScalafixImplSuite extends FunSuite with DiffAssertions {
+  def semanticdbPluginPath(): String = {
+    // Copy-pasted code from metac command-line tool
+    // https://github.com/scalameta/scalameta/blob/9f15793aae3cb6a00e1e2d6bcbf13e9c234ea91f/semanticdb/metac/src/main/scala/scala/meta/internal/metac/Main.scala#L14-L23
+    val manifestDir = Files.createTempDirectory("semanticdb-scalac_")
+    val resourceUrl =
+      classOf[SemanticdbPlugin].getResource("/scalac-plugin.xml")
+    val resourceChannel = Channels.newChannel(resourceUrl.openStream())
+    val manifestStream = new FileOutputStream(
+      manifestDir.resolve("scalac-plugin.xml").toFile)
+    manifestStream.getChannel.transferFrom(resourceChannel, 0, Long.MaxValue)
+    manifestStream.close()
+    val pluginClasspath = classOf[SemanticdbPlugin].getClassLoader match {
+      case null => manifestDir.toString
+      case cl: URLClassLoader =>
+        cl.getURLs.map(_.getFile).mkString(File.pathSeparator)
+      case cl => sys.error(s"unsupported classloader: $cl")
+    }
+    pluginClasspath
+  }
+  def scalaLibrary: AbsolutePath =
+    RuleCompiler.defaultClasspathPaths
+      .find(_.toNIO.getFileName.toString.contains("scala-library"))
+      .getOrElse {
+        throw new IllegalStateException("Unable to detect scala-library.jar")
+      }
+
+  test("versions") {
+    val api = i.Scalafix.classloadInstance(this.getClass.getClassLoader)
+    assert(api.scalafixVersion() == Versions.version)
+    assert(api.scalametaVersion() == Versions.scalameta)
+    assert(api.scala211() == Versions.scala211)
+    assert(api.scala212() == Versions.scala212)
+    assert(
+      api
+        .supportedScalaVersions()
+        .sameElements(Versions.supportedScalaVersions)
+    )
+    val help = api.mainHelp(80)
+    assert(help.contains("Usage: scalafix"))
+  }
+
+  test("error") {
+    val cl = new URLClassLoader(Array())
+    val ex = intercept[ScalafixException] {
+      i.Scalafix.classloadInstance(cl)
+    }
+    assert(ex.getCause.isInstanceOf[ClassNotFoundException])
+  }
+
+  test("runMain") {
+    // This is a full integration test that stresses the full breadth of the scalafix-interfaces API
+    val api = i.Scalafix.classloadInstance(this.getClass.getClassLoader)
+    val cwd = Files.createTempDirectory("scalafix")
+    val d = cwd.resolve("out")
+    val src = cwd.resolve("src")
+    Files.createDirectories(d)
+    Files.createDirectories(src)
+    val semicolon = src.resolve("Semicolon.scala")
+    // This rule is published to Maven Central to simplify testing --tool-classpath.
```

Comment:
I would prefer to see this in the build.

I would: add example-scalafix-rule project and make the test task depends on publishLocal of example-scalafix-rule

Reply:
I agree it would be good to bake this into the build but I opted for this because it's closer to the environment end-users will use this functionality (I've been bitten by different in custom build environments and end-user environments).

I want this integration test to be as self-contained as possible so that readers don't have to read build.sbt to understand what's going on. I can imagine this test case will be a good reference point for people who want to write unit tests when integrating with Scalafix.